File: influx_si/R/kvh.R
# 2011-03-06 sokol
# function to manipulate kvh format
obj2kvh=function(obj, objname, conct=stdout(), indent=0) {
   # write in kvh file an R object: scala, vector, matrix and list.
   # other objects are written as an output string of toString() function
   # conct is a connection opened for writing;
   # indent is tab offset for object name;
   #
   # Exemple of use:
   # m=matrix(1:6,2,3);
   # fcn=file("m.kvh", "w");
   # obj2kvh(m, "m", fcn);
   # close(fcn);
   # To add a content to existent file use "a" as open mode
   # fcn=file("m.kvh", "a");
   cls=class(obj);
   indent=max(indent,0);
   open_here=F
   if (class(conct)[1]=="character") {
      # open a file for writing if non existent yet otherwise raise an error
      if (file.exists(conct)) {
         stop(sprintf("Cannot write to existent file '%s'. To overwrite it, open a connection in 'w' or 'wb' mode.", conct))
      }
      conct=file(conct, "wb")
      open_here=T
   }
   if (is.vector(obj) && !is.list(obj) && (length(obj)==1 && is.null(names(obj)))) {
      # unnamed scalar
      cat(rep("\t", indent, sep=""), sep="", file=conct);
      cat(c(esc_kvh_k(objname), "\t", obj, "\n"), sep="", file=conct);
   } else if (cls=="matrix" || substring(cls, nchar(cls)-5)=="Matrix") {
      # matrix name
      cat(rep("\t", indent, sep=""), sep="", file=conct);
      cat(c(esc_kvh_k(objname), "\n"), sep="", file=conct);
      # place for row names
      cat(rep("\t", indent+1, sep=""), sep="", file=conct);
      cat("row_col\t", sep="", file=conct);
      # column names if any else column numbers
      if (dim(obj)[[2]] > 0) {
         if (length(dimnames(obj)[[2]])==dim(obj)[[2]]) {
            cat(dimnames(obj)[[2]], sep="\t", file=conct);
         } else {
            cat(paste("V", 1:dim(obj)[[2]], sep=""), sep="\t", file=conct);
         }
      }
      cat("\n", file=conct);
      # row name followed by the row values
      if (dim(obj)[[1]] > 0) {
         if (length(dimnames(obj)[[1]])==dim(obj)[[1]]) {
            rownms=dimnames(obj)[[1]];
         } else {
            rownms=1:dim(obj)[[1]];
         }
         for (i in 1:dim(obj)[[1]]) {
            cat(rep("\t", indent+1), sep="", file=conct);
            cat(c(rownms[i], obj[i,]), sep="\t", file=conct);
            cat("\n", file=conct);
         }
      }
   } else if (is.vector(obj) && !is.list(obj) && NROW(obj) >= 1) {
      # vector
      cat(rep("\t", indent, sep=""), sep="", file=conct);
      cat(esc_kvh_k(objname), file=conct);
      cat("\n", file=conct);
      # row name followed by the row values
      if (length(names(obj))==NROW(obj)) {
         rownms=names(obj);
      } else {
         rownms=1:NROW(obj);
      }
      for (i in 1:NROW(obj)) {
         cat(rep("\t", indent+1), sep="", file=conct);
         cat(c(rownms[i], esc_kvh_v(obj[i])), sep="\t", file=conct);
         cat("\n", file=conct);
      }
   } else if (is.list(obj)) {
      # list => recursive call if the list is not empty
      cat(rep("\t", indent, sep=""), sep="", file=conct);
      cat(esc_kvh_k(objname), file=conct);
      if (length(obj) > 0) {
         cat("\n", file=conct);
         # row name followed by the row values
         if (length(names(obj))==length(obj)) {
            rownms=names(obj);
         } else {
            rownms=1:length(obj);
         }
         for (i in 1:length(obj)) {
            obj2kvh(obj[[i]], rownms[i], conct, indent+1);
         }
      } else {
         cat("\t\n", file=conct);
      }
   } else if (length(obj)==1 && (class(obj) == "numeric" ||
         class(obj) == "character")) {
      # scalar value
      cat(rep("\t", indent, sep=""), sep="", file=conct);
      cat(c(esc_kvh_k(objname), esc_kvh_v(format(obj[[1]]))), sep="\t", file=conct);
      cat("\n", file=conct);
   } else {
      # unlnown type, write its string value
      cat(rep("\t", indent, sep=""), sep="", file=conct);
      cat(c(esc_kvh_k(objname), esc_kvh_v(format(obj))), sep="\t", file=conct);
      cat("\n", file=conct);
   }
   if (open_here) {
      close(conct)
   }
   return(NULL)
}
esc_kvh_k=function(s) {
   return(gsub("([\t\\\n])", "\\\\\\1", s));
}
esc_kvh_v=function(s) {
   return(gsub("([\\\n])", "\\\\\\1", s));
}
kvh2list=function(fp, lev=0, indent=0) {
#   Read a kvh file from fp stream descriptor
#   and organize its content in a named nested list
#   list(k1=v1, k2=list(k2.1=v2.1))
#   If at some level the first key is row_col
#   then a matrix with names row and columns is tried
#   to be read.
#   If fp is a string, it is used in open() operator
#   for binary reading
# NB: repeted key names in the same hierarchical scope will be
# silently overwritten. So only the last key=value pair will be returned
# NB: all content is of character type.


   # check the stream
   open_here=FALSE;
   if (is.character(fp)) {
      #cat(paste("open ", fp), "\n");
      oldloc=Sys.getlocale("LC_CTYPE");
      Sys.setlocale("LC_CTYPE", "en_US")
      fp=file(fp, "rb");
      seek(fp, 0);
      open_here=TRUE;
   }
   # error control
   if (lev < 0 || indent < 0) {
      stop(sprintf("lev=%d, indent=%d both must be positive", lev, indent));
   }
   if (lev < indent) {
      stop(sprintf("lev=%d, indent=%d, lev must be greater or equal to indent", lev, indent));
   }
#    don't know an equivalent of tell() in R :(
#    if lev > fp.tell():
#        raise NameError("lev=%d, file position=%d, lev must be less or equal to file position"%(lev[0], fp.tell()));
#    if indent[0] > fp.tell():
#        raise NameError("indent=%d, file position=%d, indent must be less or equal to file position"%(indent[0], fp.tell()));
   # algorithm:
   # advance to requested indent (=level)
   # if not sucsessful return an empty list
   # read a key 
   # if sep==\t read value
   # elif sep=\n
   #     recursive call
   #     if no result at the level+1 put empty value
   # else put empty value
   tlist=list();
   key="";
   val="";
   while (TRUE) {
      # make current position to point to the begining of a key
      # so go through an appropriate tab number for the current level
      while (indent < lev) {
         # set file pointer to the requested indentation
         char=readChar(fp, 1);
         if (length(char) && char!="\t") {
            if (length(char) && nchar(char)==1) {
                seek(fp,-1,origin="current");
            }
            break;
         } else if (length(char) == 0) {
            break;
         }
         indent=indent+1;
      }
      if (indent < lev) {
         # we could not read till the requested level
         # so the current level is finished;
         if (open_here) {
            Sys.setlocale("LC_CTYPE", oldloc);
            close(fp);
         }
         return(list(kvh=tlist, indent=indent));
      }
      ksv=kvh_read_keyval(fp);
      key=ksv$key;
      sep=ksv$sep;
      val=ksv$val
      if (sep=="\t") {
         #tlist[[key]]=kvh_read_val(fp);
         tlist[[key]]=val
         indent=0;
      } else if (sep=="\n") {
         lev=lev+1;
         indent=0;
         nextlev=kvh2list(fp, lev, indent);
         nextlist=nextlev$kvh;
         indent=nextlev$indent;
         if (length(nextlist) && names(nextlist)[1] == "row_col") {
            #cat("row_col list names=", names(nextlist), sep="\n");
            cnames=unlist(strsplit(nextlist[["row_col"]], "\t"));
            nc=length(cnames);
            all_col_eq=TRUE;
            m=matrix("", 0, nc);
            for (s in nextlist[-1]) {
               r=unlist(strsplit(s, "\t"));
               if (length(r) != nc) {
                  all_col_eq=FALSE;
                  break;
               }
               m=rbind(m, r);
            }
            if (all_col_eq) {
               dimnames(m)=list(names(nextlist)[-1], cnames);
               nextlist=m;
            }
         }
         lev=lev-1;
         if (length(nextlist)==0) {
             # no value and no deeper level
             tlist[[key]]="";
         } else {
             tlist[[key]]=nextlist;
         }
      } else {
         # we are at the end of file
         if (indent || nchar(key)) {
            tlist[[key]]="";
         }
         if (open_here) {
            Sys.setlocale("LC_CTYPE", oldloc);
            close(fp);
         }
         if (lev) {
            indent=0;
            return(list(kvh=tlist, indent=indent));
         } else {
            return(tlist);
         }
      }
   }
}

kvh_read_key=function(fp) {
   #  Read a string from the current position till the first unescaped
   #  \t, \n or the end of the stream fp.
   #  Return a list (key=key, sep=sep). sep="" at the end of the stream
   key="";
   while (TRUE) {
      char=readChar(fp, 1);
      if (!length(char) || !nchar(char)) {
          return(list(key=key, sep=""));
      } else if (char=="\\") {
         # try to read next char if any
         nextchar=readChar(fp, 1);
         if (!length(nextchar) || !nchar(nextchar)) {
            # end of file
            return(list(key=key,sep=""));
         } else {
            # just add the escaped char
            key=paste(key, nextchar, sep="");
         }
      } else if (char=="\t" || char=="\n") {
         return(list(key=key, sep=char));
      } else {
         # just add a plain char
         key=paste(key, char, sep="");
      }
   }
}
kvh_read_val=function(fp) {
   #  Read a string from current position till the first unescaped
   #  \n or the end of file.
   #  Return the read string.
   val="";
   while (TRUE) {
      char=readChar(fp, 1);
      if (!length(char) || !nchar(char) || char=="\n") {
         return(val);
      } else if (char=="\\") {
          # try to read next char if any
          nextchar=readChar(fp, 1);
          if (!length(nextchar) || !nchar(nextchar)) {
             # end of file
             return(val);
          } else {
             # just add escaped char
             val=paste(val, nextchar, sep="");
          }
      } else {
         # just add a plain char
         val=paste(val, char, sep="");
      }
   }
}
kvh_read_keyval=function(fp) {
   #  Read a string from the current position till the first unescaped
   #  \n or the end of the stream fp.
   #  Return a list (key=key, sep=sep, val=val). sep="" at the end of the stream
   totrow=""
   while (TRUE) { # read till the end of line is unescaped \n or fp end
      row=readLines(fp, 1, warn=F)
      if (length(row)==0 && nchar(totrow)==0) {
         # we were at the end
         return(list(key="", val="", sep=""))
      }
      totrow=paste(totrow, row, sep="", collapse="")
      # see if we need to read further
      rend=regexpr("\\\\*$", row)
#print(paste("row=", row, "rend=", rend))
      if (attr(rend, "match.length") %% 2 == 0) {
         # the row end is not escaped => it is completely read
         break
      } else {
         # append the next row
         next
      }
   }
   # split the totrow in key, sep, val
   rsep=gregexpr("[\\]*\t", totrow)
   isep=attr(rsep[[1]], "match.length")
   isep=which(isep != -1 && isep %% 2 == 1)
   if (length(isep)) {
      # there is a separator tab
      endkey=rsep[[1]][isep[1]]+attr(rsep[[1]], "match.length")[isep[1]]-2
      key=substring(totrow, 1, endkey)
      sep="\t"
      val=substring(totrow, endkey+2)
   } else {
      # no tab separator
      key=totrow
      sep="\n"
      val=NULL
   }
   # strip backslashes
   keyval=as.list(gsub("\\\\(.)", "\\1", c(key=key, val=val)))
   keyval$sep=sep
   return(keyval)
}
kvh_get_matrix=function(f, v) {
   # return a plain matrix stored in the field v of a kvh file f.
   # v can be a vector of subfield keys.
   
   # read the file in a vector
   cont=readLines(f)
   ncont=length(cont)
   # get start line number by grep successively all fields in v
   # and the indent
   indent=0
   nstart=1
   for (it in v) {
      #res=system(sprintf("tail -n +%d '%s' 2>&1| grep -m 1 -n '\\t*%s$'", nstart+1, f, it), intern=T, ignore.stderr=T)
      nstart=nstart+grep(sprintf("\\t*\\Q%s\\E$", it), cont[nstart:ncont])[1]
      indent=which(nchar(strsplit(cont[nstart], "\t", fix=T)[[1]])>0)[1]-1
   }
   # get end number of the matrix in the kvh
   #res=system(sprintf("tail -n +%d '%s' 2>&1| grep -m 1 -n -E ^'%s\\w'", nstart+1, f, join("", rep("\t", indent))), intern=T)
   nend=grep(sprintf("^\\t{0,%d}[^\t]", indent-1), cont[nstart:ncont])
   if (length(nend)) {
      nend=nstart+nend[1]-2
   } else {
      nend=ncont
   }
   d=matrix(unlist(strsplit(cont[nstart:nend], "\t")), byrow=T, nrow=nend-nstart+1)[,-(1:indent), drop=F]
   rownames(d)=d[,1]
   if (rownames(d)[1]=="row_col") {
      colnames(d)=d[1,]
      d=d[-1,-1,drop=F]
   } else {
      d=d[,-1,drop=F]
   }
   dn=dimnames(d)
   wop=options()$warn
   options(warn=-1)
   num<-as.numeric(d)
   options(warn=wop)
   if (!any(is.na(num))) {
      d=matrix(num, nrow=nrow(d))
      dimnames(d)=dn
   }
   return(d)
}

File: influx_si/kvh.py
import sys
import codecs
from re import match

#import pdb

def isstr(s):
    return isinstance(s, str)
class Obj():
    def __init__(**kwargs):
        self.__dict__.update(kwrags)
def kvh2tlist(fp, lev=[0], indent=[0], strip=False):
    """
    Read a kvh file from fp stream descriptor
    and organize its content in list of tuples [(k1,v1), (k2,[(k2.1, v2.1)])]
    If fp is a string, it is used in open() operator
    """
    # check the stream
    open_here=False;
    if isstr(fp):
        fp=codecs.open(fp, "r", encoding="utf-8-sig");
        fp.seek(0);
        open_here=True;
    # error control
    if lev[0] < 0 or indent[0] < 0:
        raise NameError("lev=%d, indent=%d both must be positive"%(lev[0], indent[0]));
    if lev[0] < indent[0]:
        raise NameError("lev=%d, indent=%d, lev must be greater or equal to indent"%(lev[0], indent[0]));
    if fp != sys.stdin and lev[0] > fp.tell():
        raise NameError("lev=%d, file position=%d, lev must be less or equal to file position"%(lev[0], fp.tell()));
    if fp != sys.stdin and indent[0] > fp.tell():
        raise NameError("indent=%d, file position=%d, indent must be less or equal to file position"%(indent[0], fp.tell()));
    # algorithm:
    # advance to requested indent (=level)
    # if not sucsessful return an empty list
    # read a key 
    # if sep==\t read value
    # elif sep==\n
    #     recursive call with increased indentation
    #     if no result at the level+1 put empty value
    # else put empty value
    tlist=[];
    key="";
    val="";
    while True:
        # current position is supposed to point to the begining of a key
        # so go through an appropriate tab number for the current level
        while indent[0] < lev[0]:
            char=fp.read(1);
            if char!="\t":
                if char!="":
                    fp.seek(-1,1);
                break;
            indent[0]+=1;
        if indent[0] < lev[0]:
            # we could not read till the requested level
            # so the current level is finished;
            if open_here:
                fp.close();
            return tlist;
        (key,sep)=kvh_read_key(fp, strip);
        if sep=="\t":
            tlist.append((key, kvh_read_val(fp, strip)));
            indent[0]=0;
        elif sep=="\n":
            lev[0]+=1;
            indent[0]=0;
            nextlist=kvh2tlist(fp, lev, indent, strip);
            lev[0]-=1;
            if len(nextlist)==0:
                # no value and no deeper level
                tlist.append((key, ""));
            else:
                tlist.append((key, nextlist));
        else:
            # we are at the end of file
            if indent[0] or key:
                tlist.append((key, ""));
            indent[0]=0;
            lev[0]=0;
            if open_here:
                fp.close();
            return tlist;

def kvh_read_key(fp, strip=False):
    """Read a string from the current position till the first unescaped \t, \n or the end of stream fp.

:returns: tuple (key, sep), sep=None at the end of the stream

"""
    #pdb.set_trace();##
    key="";
    while True:
        char=fp.read(1);
        if char=="\\":
            # try to read next char if any
            nextchar=fp.read(1);
            if nextchar=="":
                # end of file
                return (key.strip() if strip else key, None);
            else:
                # just add escaped char
                key+=nextchar;
        elif char=="\t" or char=="\n":
            return (key.strip() if strip else key, char);
        elif char=="":
            return (key.strip() if strip else key, None);
        else:
            # just add a plain char
            key+=char;
def kvh_read_val(fp, strip=False):
    """
    Read a string from current position till the first unescaped
    \n or the end of file.
    Return the read string."""
    val="";
    while True:
        char=fp.read(1);
        if char=="\\":
            # try to read next char if any
            nextchar=fp.read(1);
            if nextchar=="":
                # end of file
                return val.strip() if strip else val;
            else:
                # just add escaped char
                val+=nextchar;
        elif char=="\n" or char=="":
            return val.strip() if strip else val;
        else:
            # just add a plain char
            val+=char;
def kvh_tlist2dict(tlist):
    """
    Translate a tlist structure read from a kvh file to
    a hierarchical dictionnary. Repeated keys at the same level
    of a dictionnary are silently overwritten"""
    return dict((k,(v if isstr(v) else kvh_tlist2dict(v))) for (k,v) in tlist);
def kvh_tlist2obj(tlist):
    """
    Translate a tlist structure read from a kvh file to
    a hierarchical dictionnary. Repeated keys at the same level
    of a dictionnary are silently overwritten"""
    return Obj(**dict((k,(v if isstr(v) else kvh_tlist2obj(v))) for (k,v) in tlist));
def kvh2dict(fp, strip=False):
    r"""
    Read a kvh file from fp pointer then translate its tlist
    structure to a returned hierarchical dictionnary.
    Repeated keys at the same level of a dictionnary are
    silently overwritten"""
    return kvh_tlist2dict(kvh2tlist(fp, strip=strip));
def kvh2obj(fp, strip=False):
    r"""
    Read a kvh file from fp pointer then translate its tlist
    structure to a returned object hierarchy.
    Repeated fields at the same level of an object are
    silently overwritten"""
    return kvh_tlist2obj(kvh2tlist(fp, strip=strip));
def dict2kvh(d, fp=sys.stdout, indent=0):
    r"""dict2kvh(d, fp=sys.stdout, indent=0)
    Write a nested dictionary on the stream fp (stdout by default).
    """
    open_here=False;
    if isstr(fp):
        open_here=True;
        fp=open(fp, "w");
    for (k,v) in d.items():
        fp.write("%s%s" % ("\t"*indent, escape(str(k), "\t\\\n")));
        if type(v) == type({}) or type({}) in type(v).__bases__:
            # recursive call with incremented indentation
            fp.write("\n");
            dict2kvh(v, fp, indent+1);
        elif "__dict__" in dir(v) and v.__dict__:
            # recursive call with incremented indentation
            fp.write("\n");
            dict2kvh(v.__dict__, fp, indent+1);
        else:
            fp.write("\t%s\n" % escape(str(v), "\\\n"));
    if open_here:
        fp.close();
def tlist2kvh(d, fp=sys.stdout, indent=0):
    r"""tlist2kvh(d, fp=sys.stdout, indent=0)
    Write a (hierarchichal) list of 2-tuples on the stream fp (stdout by default).
    """
    open_here=False;
    if isstr(fp):
        open_here=True;
        fp=open(fp, "w");
    for (k,v) in d:
        fp.write("%s%s" % ("\t"*indent, escape(str(k), "\t\\\n")));
        if type(v) == type([]):
            # recursive call with incremented indentation
            fp.write("\n");
            tlist2kvh(v, fp, indent+1);
        else:
            fp.write("\t%s\n" % escape(str(v), "\\\n"));
    if open_here:
        fp.close();
def kvh_getv_by_k(kvt, kl):
    r"""kvh_getv_by_k(kvt, kl)->None|String|kvh tlist
    get value from kvt (kvh tlist) according to the key hierarchy
    defined in the list of keys kl. Return None if no key is found
    """
    for (k,v) in kvt:
        if k==kl[0]:
            # found
            if len(kl) == 1:
                return(v);
            elif len(kl) > 1:
                # recursive call
                return(kvh_getv_by_k(v, kl[1:]));
def escape(s, spch="|&;<>()$`\\\"' \t\n*?[#~=%", ech="\\"):
    r"""escape(s, spch="|&;<>()$`\\\"' \t\n*?[#~=%", ech="\\")
escape special characters in s. The special characters are listed in spch.
Escaping is done by putting an ech string before them.
Default spch and ech corresponds to quoting Shell arguments
in accordance with
http://www.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html
Example: os.system("ls %s" % escape(file_name_with_all_meta_chars_but_newline));
.. note:

1. Escaped <newline> is removed by a shell if not put in a single-quotted string (' ')

2. A single-quote character even escaped cannot appear in a single-quotted string
"""
    return "".join((ech+c if c in spch else c) for c in s);

def kvh_get_matrix(fp, keys):
    """Get matrix or vector whose key suite is in a list keys from a kvh file given in fp
    (file pointer of file name). For big kvh files, this function can be much faster
    than kvh2tlist()+kvh_getv_by_k()
    Return a matrix which is a list of lists (rows). The first item in each row is the
    row name. In case of matrix (i.e. "row_col" is present in kvh file), the very first row contain column names."""
    
    if isinstance(fp, str):
        with open(fp, "r") as fp:
            cont=fp.readlines()
    else:
        cont=fp.readlines()
    # keys can be a list of subfield keys.
    ncont=len(cont)
    # get start line number by grep successively all fields in v
    # and the indent
    indent=0
    nstart=0
    #pdb.set_trace()
    for k in keys:
        for i in range(nstart, ncont):
            if match(r"^\t{%d,}%s\r?$"%(indent,escape(k)), cont[i]):
                nstart=i+1
                o=match(r"^\t*", cont[i])
                indent=len(o.group(0))
                break
        else:
            raise NameError("The key '%s' was not found in kvh file '%s'."%(k, fp.name))
    # get end number of the matrix row in the kvh
    nend=ncont
    for i in range(nstart, ncont):
        if match(r"^\t{0,%d}[^\t]"%indent, cont[i]):
            nend=i
            break
    # get matrix
    d=[s[(indent+1):].rstrip("\r\n").split("\t") for s in cont[nstart:nend]]
    return d
